src/server/bg_services/semapore_monitor.js

```diff
@@ -0,0 +1,103 @@
+/* Copyright (C) 2016 NooBaa */
+'use strict';
+
+const dbg = require('../../util/debug_module')(__filename);
+const config = require('../../../config');
+const endpoint_stats_collector = require('../../sdk/endpoint_stats_collector').instance();
+const { buffers_pool } = require('../../sdk/namespace_fs');
+
+class SemaphoreMonitor {
+
+    /**
+     * @param {{
+     *   name: string;
+     *   object_io: import('../../sdk/object_io'),
+     *   report_bulk_sizes: Array<Object>,
+     *   should_monitor: boolean;
+     * }} params
+     */
+    constructor({ name, object_io, report_bulk_sizes, should_monitor }) {
+        this.name = name;
+        this.object_io = object_io;
+        this.report_bulk_sizes = report_bulk_sizes;
+        this.should_monitor = should_monitor;
+        console.log(name, object_io, report_bulk_sizes, should_monitor)
+    }
+
+    async run_batch() {
+        dbg.log1("semaphore_monitor: START");
+        if (!this._can_run()) return;
+        try {
+            await this.run_semaphore_monitor();
+        } catch (err) {
+            dbg.error('semaphore_monitor:', err, err.stack);
+        }
+        return config.SEMAPHORE_MONITOR_DELAY;
+    }
+
+    async run_semaphore_monitor() {
+        if (!this.should_monitor) return;
+        try {
+            await this.start_semaphore_monitor(this.object_io);
+            await this.start_nsfs_semaphore_monitor();
+        } catch (err) {
+            dbg.error('semaphore_monitor:', err, err.stack);
+        }
+    }
+
+    _can_run() {
+        if (!this.object_io) {
+            dbg.log0('semaphore_monitor: object_io is invalid', this.object_io);
+            return false;
+        }
+        if (!this.report_bulk_sizes || this.report_bulk_sizes.length === 0) {
+            dbg.log0('semaphore_monitor: report_bulk_sizes do not have valid size', this.report_bulk_sizes);
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * @param {import('../../sdk/object_io')} io_buffers
+     */
+    async start_semaphore_monitor(io_buffers) {
+        try {
+            const semaphore_report = {
+                timestamp: Date.now(),
+                semaphore_state: {
+                    semaphore_cap: config.IO_SEMAPHORE_CAP,
+                    value: io_buffers._io_buffers_sem.value,
+                    waiting_value: io_buffers._io_buffers_sem.waiting_value,
+                    waiting_time: io_buffers._io_buffers_sem.waiting_time,
+                    waiting_queue: io_buffers._io_buffers_sem._wq.length,
+                }
+            };
+            await endpoint_stats_collector.update_semaphore_state(semaphore_report, "ns", this.report_bulk_sizes);
```

Comment:
This function shouldn't be `async`, and there is no need to `await` here, since `update_semaphore_state` is not async